### PR TITLE
Changing create fs to NOT do a head operation before writing

### PR DESCRIPTION
**Pull request by LarryFinn**
This causes eventual consistency, which is esp stupid for our use case.

Warning, I am not sure if this will break everything 👀

### CODE REVIEW

**Comment by LarryFinn** [comment]
I think this is a relatively safe change because in master this function looks like
```
  public FSDataOutputStream create(Path f, FsPermission permission, boolean overwrite, 
    int bufferSize, short replication, long blockSize, Progressable progress) throws IOException {
    String key = pathToKey(f);

    if (!overwrite && exists(f)) {
      throw new FileAlreadyExistsException(f + " already exists");
    }
    if (getConf().getBoolean(FAST_UPLOAD, DEFAULT_FAST_UPLOAD)) {
      return new FSDataOutputStream(new S3AFastOutputStream(s3, this, bucket,
          key, progress, statistics, cannedACL,
          serverSideEncryptionAlgorithm, partSize, multiPartThreshold,
          threadPoolExecutor), statistics);
    }
    // We pass null to FSDataOutputStream so it won't count writes that are being buffered to a file
    return new FSDataOutputStream(new S3AOutputStream(getConf(), transfers, this,
      bucket, key, progress, cannedACL, statistics, 
      serverSideEncryptionAlgorithm), null);
  }
```

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java [hunk 1]
```diff
@@ -623,23 +623,21 @@ public FSDataOutputStream create(Path f, FsPermission permission,
...
-      if (status.isDirectory()) {
-        // path references a directory: automatic error
-        throw new FileAlreadyExistsException(f + " is a directory");
-      }
-      if (!overwrite) {
-        // path references a file and overwrite is disabled
+        // if the thread reaches here, there is something at the path
+        if (status.isDirectory()) {
+          // path references a directory: automatic error
+          throw new FileAlreadyExistsException(f + " is a directory");
```
Comment (MasterDDT) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: In the old behavior, if you overwrite and its a directory, its supposed to throw

In new behavior, if you overwrite and its a directory, its just gonna overwrite?

**Comment by MasterDDT** [comment]
Whats the root bug here? If you're overwriting something in s3...my recommendation would be just dont. Write new files and then have some row in SQL be a pointer to it, and overwrite that.

**Comment by LarryFinn** [comment]
@MasterDDT the problem here is when we are uploading files (typically for the first time), hadoop fs abstraction does a HEAD operation first to prevent you from overwriting if you didn't want to overwrite.  which is not something we care about.  Here is the blurb from amazon
```
Amazon S3 Data Consistency Model

Amazon S3 provides read-after-write consistency for PUTS of new objects in your S3 bucket in all Regions with one caveat. The caveat is that if you make a HEAD or GET request to the key name (to find if the object exists) before creating the object, Amazon S3 provides eventual consistency for read-after-write. 
```

**Comment by MasterDDT** [comment]
> @MasterDDT the problem here is when we are uploading files (typically for the first time), hadoop fs abstraction does a HEAD operation first to prevent you from overwriting if you didn't want to overwrite. which is not something we care about. Here is the blurb from amazon
> 
> Amazon S3 Data Consistency Model
> 
> Amazon S3 provides read-after-write consistency for PUTS of new objects in your S3 bucket in all Regions with one caveat. The caveat is that if you make a HEAD or GET request to the key name (to find if the object exists) before creating the object, Amazon S3 provides eventual consistency for read-after-write.

Do you have the callstack to the `S3AFileSystem.create`? I wonder if its just a matter of not setting `overwrite` [here](https://github.com/ActionIQ/aiq/blob/49e8e77299b7309174cf4409681e11ba135ba120/src/scala/co/actioniq/flame/SparkImplicits.scala#L336) (bc it should be a new file anyways), then it should not pass overwrite down I think.

The other case, where you cant read something you just wrote (and you never HEAD/GET it before), we solve by [doing retries](https://github.com/ActionIQ/aiq/blob/49e8e77299b7309174cf4409681e11ba135ba120/src/scala/co/actioniq/flame/SparkImplicits.scala#L156)

**Comment by LarryFinn** [comment]
Im pretty sure we don't care about overwriting.  

Yeah, we retry for absolutely no reason.  Why waste so many cycles on retrying?

**Comment by MasterDDT** [comment]
> Im pretty sure we don't care about overwriting.
> 
> Yeah, we retry for absolutely no reason. Why waste so many cycles on retrying?

Oh now I remember, for spark at least we need to do `parquetWrite(mode=overwrite)` because if it retries a task, the file may already exist (and it will write it with the same exact name).

What exactly is the callstack to your failure? Is it a spark write or direct S3 API write?

**Comment by MasterDDT** [comment]
[HADOOP-16490](https://issues.apache.org/jira/browse/HADOOP-16490) looks like the fix you want, but it requires s3guard stuff 🤔

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java (hunk 1 above)
Comment (MasterDDT) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: I guess if it fails with overwrite=true, or overwrites the directory...either case seems fine to me.

**Comment by LarryFinn** [comment]
@MasterDDT tests ran fine but i should add something
not sure how we deploy spark... how would we tell it to grab from our artifacts repo in s3?

**Comment by MasterDDT** [comment]
> @MasterDDT tests ran fine but i should add something
> not sure how we deploy spark... how would we tell it to grab from our artifacts repo in s3?

Its just configuring maven to look in the s3 bucket, shouldnt be much diff than how we tell ivy to do that.

Btw...still waiting on the actual root error here. Why isnt retrying good enough? We've got lots of other places we write to s3 and workaround this, why is this case special?

**Comment by LarryFinn** [comment]
@MasterDDT  there is no root error.  This method doing head before writing to s3 causes us to HAVE to do retry.  It forces us into eventual consistency when it doesn't add any value

**Comment by MasterDDT** [comment]
> @MasterDDT there is no root error. This method doing head before writing to s3 causes us to HAVE to do retry. It forces us into eventual consistency when it doesn't add any value. Why do many retries because of eventual consistency when we do not have to?

Yep I got that, I meant whats the AIQ jira that is causing us to change this?

**Comment by LarryFinn** [comment]
We saw we were getting a lot of failed exports because of a bad refactoring of the retry code path in ftpuploader when grabbing the file from s3.  Takis fixed that bug, but I was thinking why do we even have all this eventual consistency with s3 when we are uploading a file once.  I can create a jira?

**Comment by MasterDDT** [comment]
I see, so its now `SparkImplicits.parquetRead`, because that already does a retry. Yep open a jira, just so we know exactly where the s3 read is happening.

Btw to be clear, HEAD then write then GET isn't *guaranteed* to fail right? It just means if you come back too fast, they wont have invalidate the 404 response. Is there a reason we write a file and then immediate read it back (as opposed to just keep it in say spark mem or something)?

But yeah I think this fix is fine. I thought/hope we have metrics on how often Retries happen, so we should see a drop right?

**Comment by LarryFinn** [comment]
Yeah, it's not guaranteed to fail but I'm thinking it does fairly often. I think this is happens between apps, like when qapi writes parquet and dp reads it or dp writes csv and exporting reads it. We have metrics on 4xx responses from s3, but Im going to make sure we have the right retry metrics before moving forward.

**Comment by MasterDDT** [comment]
P.S you prob also want to change the version in the pom, `2.9.2` -> `2.9.2-aiq1` or something. And then build the jar, put it up in `s3://aiq-artifacts/releases/x/y/z`, and then we can figure out how to get spark to pull it down

**Pull request merged**